routes/open-id.ts

```diff
@@ -0,0 +1,201 @@
+
+  var User = null; // Will be set by initUserModel
+
+  var sm = stoutmeal(stoutmealDefaults(app.config.stoutmeal));
+
+  var controller = Controller();
+  controller.app.set('view engine', 'jade');
+
+  controller.define('index', function (req, res) {
+    res.render('open-id/index', app.settings);
+  });
+
+  controller.define('success', function (req, res) {
+    res.send(200, 'Successfully logged in');
+  });
+
+  function initUserModel(done) {
+    if (User) return done(null, User);
+    var seraph = seraphInit(app.config.neo4j);
+    barleyInit(seraph, {}, function (err, models) {
+      if (err) {
+        app.log.error('failed to initiale models. error follows:');
+        app.log.error(err.message);
+        app.log.error(err);
+        return done('Open id plugin failed to initialize User model.');
+      }
+      User = models.user;
+      done(null, User);
+    });
+  }
+
+  // Return falsy if access allowed, or error string if disallowed
+  function accessDisallowed(openIDUser: IOpenIDUser) {
+    if (!openIDUser.sub) {
+      return 'ID is required';
+    }
+    if (!openIDUser.email) {
+      return 'Email is required';
+    }
+    return false;
+  }
+
+  function invalidateAuthKey(req, res) {
+    var authKey = sm.cookie.get(req);
+    if (authKey) {
+      sm.auth.invalidateAuthKey(authKey, function (err) {
+        if (err) {
+          app.log.error("Could not invalidate user's auth key");
+          app.log.error(err);
+        } else {
+          app.log.info('invalidated authKey ' + authKey);
+        }
+        logoutCompleted(res);
+      });
+    } else {
+      logoutCompleted(res);
+    }
+  }
+
+  function logoutCompleted(res) {
+    res.redirect(defaultRedirect);
+  }
+
+  const passportAuth = passport.authenticate('open-id', {
+    failureRedirect: '/integration/open-id/failure',
+    failureFlash: true,
+  });
+
+  controller.define('login', function (req, res) {
+    var reqUrl = url.parse(req.url, true);
+    var redirect = reqUrl.query['redirect'];
+    res.cookie(REDIRECT_COOKIE, redirect || '/', { maxAge: 60 * 60 * 1000 });
+
+    passportAuth(req, res);
+  });
+
+  controller.define('loginCallback', function (req, res) {
+    function locallyAuthenticateOpenIdUser(openIDUser: IOpenIDUser) {
+      app.log.info(
+        'Open ID auth login callback success: ' + openIDUser && openIDUser.sub
+      );
+
+      var reason = accessDisallowed(openIDUser);
+      if (reason) {
+        res.render('open-id/accessRequirementFailed', {
+          reason: reason,
+          userName: req.user.name || req.user.sub || 'Your Open ID user',
+          json: {
+            reason: JSON.stringify(reason),
+            user: JSON.stringify(req.user),
+          },
+        });
+        return;
+      }
+
+      const openIDUserId = openIDUser.sub;
+      const translatedUser = {
+        name: openIDUser.name || openIDUserId,
+        email: openIDUser.email || '',
+      };
+
+      async.waterfall(
```

Comment:
Haha, oh man. Welcome to our antique codebase where we have this relic of the distant past `async.waterfall` instead of async await 😬

Reply:
😆 Yeah.. I was going back and forth whether I should rewrite to use async await. Figured it was too time consuming 🤔 Might revisit later though 👍